#📌 Extracão

In [1]:
import requests
import pandas as pd
import json

In [2]:
url = 'https://raw.githubusercontent.com/Enoque12/Analise_TelecomX_Challenge2/refs/heads/main/Dados_API/TelecomX_Data.json'

In [3]:
dados = requests.get(url)

In [4]:
resultados = json.loads(dados.text)

In [5]:
df = pd.json_normalize(resultados)

In [6]:
df

customerID Churn customer.gender  customer.SeniorCitizen  \
0     0002-ORFBO    No          Female                       0   
1     0003-MKNFE    No            Male                       0   
2     0004-TLHLJ   Yes            Male                       0   
3     0011-IGKFF   Yes            Male                       1   
4     0013-EXCHZ   Yes          Female                       1   
...          ...   ...             ...                     ...   
7262  9987-LUTYD    No          Female                       0   
7263  9992-RRAMN   Yes            Male                       0   
7264  9992-UJOEL    No            Male                       0   
7265  9993-LHIEB    No            Male                       0   
7266  9995-HOTOH    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
0                 Yes                 Yes                9                Yes   
1                  No                  No                9                Yes   
2                  No                  No                4                Yes   
3                 Yes                  No               13                Yes   
4                 Yes                  No                3                Yes   
...               ...                 ...              ...                ...   
7262               No                  No               13                Yes   
7263              Yes                  No               22                Yes   
7264               No                  No                2                Yes   
7265              Yes                 Yes               67                Yes   
7266              Yes                 Yes               63                 No   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
0                     No                      DSL  ...                   Yes   
1                    Yes                      DSL  ...                    No   
2                     No              Fiber optic  ...                    No   
3                     No              Fiber optic  ...                   Yes   
4                     No              Fiber optic  ...                    No   
...                  ...                      ...  ...                   ...   
7262                  No                      DSL  ...                    No   
7263                 Yes              Fiber optic  ...                    No   
7264                  No                      DSL  ...                   Yes   
7265                  No                      DSL  ...                    No   
7266    No phone service                      DSL  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                           No                  Yes                  Yes   
1                           No                   No                   No   
2                          Yes                   No                   No   
3                          Yes                   No                  Yes   
4                           No                  Yes                  Yes   
...                        ...                  ...                  ...   
7262                        No                  Yes                   No   
7263                        No                   No                   No   
7264                        No                   No                   No   
7265                       Yes                  Yes                   No   
7266                       Yes                   No                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
0                          No         One year                      Yes   
1                         Yes   Month-to-month                       No   
2                          No   Month-to-month                      Yes   
3                         Yes   Month-to-month                    

In [7]:
colunas = df.columns
nomeColunasactual = ['customerID', 'Churn','gender', 'SeniorCitizen','Partner',
                     'Dependents', 'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
                     'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV',
                     'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'Charges.Monthly', 'Charges.Total']

In [8]:
for i in range(len(colunas)):
  df = df.rename(columns={colunas[i]:nomeColunasactual[i]})

In [9]:
df

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0     0002-ORFBO    No  Female              0     Yes        Yes       9   
1     0003-MKNFE    No    Male              0      No         No       9   
2     0004-TLHLJ   Yes    Male              0      No         No       4   
3     0011-IGKFF   Yes    Male              1     Yes         No      13   
4     0013-EXCHZ   Yes  Female              1     Yes         No       3   
...          ...   ...     ...            ...     ...        ...     ...   
7262  9987-LUTYD    No  Female              0      No         No      13   
7263  9992-RRAMN   Yes    Male              0     Yes         No      22   
7264  9992-UJOEL    No    Male              0      No         No       2   
7265  9993-LHIEB    No    Male              0     Yes        Yes      67   
7266  9995-HOTOH    No    Male              0     Yes        Yes      63   

     PhoneService     MultipleLines InternetService  ... OnlineBackup  \
0             Yes                No             DSL  ...          Yes   
1             Yes               Yes             DSL  ...           No   
2             Yes                No     Fiber optic  ...           No   
3             Yes                No     Fiber optic  ...          Yes   
4             Yes                No     Fiber optic  ...           No   
...           ...               ...             ...  ...          ...   
7262          Yes                No             DSL  ...           No   
7263          Yes               Yes     Fiber optic  ...           No   
7264          Yes                No             DSL  ...          Yes   
7265          Yes                No             DSL  ...           No   
7266           No  No phone service             DSL  ...          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No         Yes         Yes              No        One year   
1                  No          No          No             Yes  Month-to-month   
2                 Yes          No          No              No  Month-to-month   
3                 Yes          No         Yes             Yes  Month-to-month   
4                  No         Yes         Yes              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7262               No         Yes          No              No        One year   
7263               No          No          No             Yes  Month-to-month   
7264               No          No          No              No  Month-to-month   
7265              Yes         Yes          No             Yes        Two year   
7266              Yes          No         Yes             Yes        Two year   

     PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0                 Yes      Mailed check           65.60          593.3  
1                  No      Mailed check           59.90          542.4  
2                 Yes  Electronic check           73.90         280.85  
3                 Yes  Electronic check           98.00        1237.85  
4                 Yes      Mailed check           83.90          267.4  
...               ...               ...             ...            ...  
7262               No      Mailed check           55.15          742.9  
7263              Yes  Electronic check           85.10         1873.7  
7264              Yes      Mailed check           50.30          92.75  
7265               No      Mailed check           67.85        4627.65  
7266               No  Electronic check           59.00         3707.6  

[7267 rows x 21 columns]

#🔧 Transformação

##**Conhecendo o DataSet**

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


Acções necessárias:
- Verificar e remover duplicados
- Valores vazios
- Nulos
- Tipos de Dados
- Valores atipícos
- Consistencia e validez de dados categoricos
- Normalização de Textos

Verificar e remover duplicados

In [11]:
print(f'A quantidade de valores duplicados no DF é: {df.duplicated().sum()}')

A quantidade de valores duplicados no DF é: 0


Consistencia e validez de dados categoricos

In [12]:
colunas = df.columns

for col in colunas:
  print('---------------------------------------------')
  print(f'Os valores únicos da Coluna "{col}": \n{df[col].unique()}')
  print('---------------------------------------------')

---------------------------------------------
Os valores únicos da Coluna "customerID": 
['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']
---------------------------------------------
---------------------------------------------
Os valores únicos da Coluna "Churn": 
['No' 'Yes' '']
---------------------------------------------
---------------------------------------------
Os valores únicos da Coluna "gender": 
['Female' 'Male']
---------------------------------------------
---------------------------------------------
Os valores únicos da Coluna "SeniorCitizen": 
[0 1]
---------------------------------------------
---------------------------------------------
Os valores únicos da Coluna "Partner": 
['Yes' 'No']
---------------------------------------------
---------------------------------------------
Os valores únicos da Coluna "Dependents": 
['Yes' 'No']
---------------------------------------------
---------------------------------------------
Os

Somatório dos Vazios e nulos

In [13]:
for col in colunas:
  quantVazio = 0
  for valor in df[col]:
    if str(valor).strip() == '':
      quantVazio += 1
  if quantVazio > 0:
    print(f'A quantidade de valores vazios na Coluna "{col}" é: {quantVazio}')


A quantidade de valores vazios na Coluna "Churn" é: 224
A quantidade de valores vazios na Coluna "Charges.Total" é: 11


Tornando os dados consistentes e normalizados

In [14]:
df_novo = df.copy()

In [15]:
df_novo['Churn'] = df_novo['Churn'].str.replace('NoNNooNo', 'No')

In [16]:
df_novo[df_novo['Charges.Total'].str.strip() == '']

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
975   1371-DWPAZ    No  Female              0     Yes        Yes       0   
1775  2520-SGTTA    No  Female              0     Yes        Yes       0   
1955  2775-SEFEE    No    Male              0      No        Yes       0   
2075  2923-ARZLG    No    Male              0     Yes        Yes       0   
2232  3115-CZMZD    No    Male              0      No        Yes       0   
2308  3213-VVOLG    No    Male              0     Yes        Yes       0   
2930  4075-WKNIU    No  Female              0     Yes        Yes       0   
3134  4367-NUYAO    No    Male              0     Yes        Yes       0   
3203  4472-LVYGI    No  Female              0     Yes        Yes       0   
4169  5709-LVOEQ    No  Female              0     Yes        Yes       0   
5599  7644-OMVMY    No    Male              0     Yes        Yes       0   

     PhoneService     MultipleLines InternetService  ...         OnlineBackup  \
975            No  No phone service             DSL  ...                  Yes   
1775          Yes                No              No  ...  No internet service   
1955          Yes               Yes             DSL  ...                  Yes   
2075          Yes                No              No  ...  No internet service   
2232          Yes                No              No  ...  No internet service   
2308          Yes               Yes              No  ...  No internet service   
2930          Yes               Yes             DSL  ...                  Yes   
3134          Yes               Yes              No  ...  No internet service   
3203           No  No phone service             DSL  ...                   No   
4169          Yes                No             DSL  ...                  Yes   
5599          Yes                No              No  ...  No internet service   

         DeviceProtection          TechSupport          StreamingTV  \
975                   Yes                  Yes                  Yes   
1775  No internet service  No internet service  No internet service   
1955                   No                  Yes                   No   
2075  No internet service  No internet service  No internet service   
2232  No internet service  No internet service  No internet service   
2308  No internet service  No internet service  No internet service   
2930                  Yes                  Yes                  Yes   
3134  No internet service  No internet service  No internet service   
3203                  Yes                  Yes                  Yes   
4169                  Yes                   No                  Yes   
5599  No internet service  No internet service  No internet service   

          StreamingMovies  Contract PaperlessBilling  \
975                    No  Two year               No   
1775  No internet service  Two year               No   
1955                   No  Two year              Yes   
2075  No internet service  One year              Yes   
2232  No internet service  Two year               No   
2308  No internet service  Two year               No   
2930                   No  Two year               No   
3134  No internet service  Two year               No   
3203                   No  Two year              Yes   
4169                  Yes  Two year               No   
5599  No internet service  Two year               No   

                  PaymentMethod Charges.Monthly  Charges.Total  
975     Credit card (automatic)           56.05                 
1775               Mailed check           20.00                 
1955  Bank transfer (automatic)           61.90                 
2075               Mailed check           19.70                 
2232               Mailed check           20.25                 
2308               Mailed check           25.35                 
2930               Mailed check           73.35                 
3134               Mailed check           25.75                 
3203  Ban

In [17]:
for i in range(len(df_novo['Charges.Total'])):
    if str(df_novo['Charges.Total'].iloc[i]).strip() == '':
      df_novo['Charges.Total'].iloc[i] = 0

/tmp/ipython-input-17-2587400656.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_novo['Charges.Total'].iloc[i] = 0
/tmp/ipython-input-17-2587400656.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_novo['Charges.Total'].iloc[i] = 0
/tmp/ipython-input-17-2587400656.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_novo['Charges.Total'].iloc[i] = 0
/tmp/ipython-input-17-2587400656.py:3: SettingWithCopyWarning: 
A value i

In [18]:
df_novo['MultipleLines'] = df_novo['MultipleLines'].str.replace('No phone service', 'No')
df_novo['OnlineSecurity'] = df_novo['OnlineSecurity'].str.replace('No internet service', 'No')
df_novo['OnlineBackup'] = df_novo['OnlineBackup'].str.replace('No internet service', 'No')
df_novo['DeviceProtection'] = df_novo['DeviceProtection'].str.replace('No internet service', 'No')
df_novo['TechSupport'] = df_novo['TechSupport'].str.replace('No internet service', 'No')
df_novo['StreamingTV'] = df_novo['StreamingTV'].str.replace('No internet service', 'No')
df_novo['StreamingMovies'] = df_novo['StreamingMovies'].str.replace('No internet service', 'No')

In [19]:
df_novo

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0     0002-ORFBO    No  Female              0     Yes        Yes       9   
1     0003-MKNFE    No    Male              0      No         No       9   
2     0004-TLHLJ   Yes    Male              0      No         No       4   
3     0011-IGKFF   Yes    Male              1     Yes         No      13   
4     0013-EXCHZ   Yes  Female              1     Yes         No       3   
...          ...   ...     ...            ...     ...        ...     ...   
7262  9987-LUTYD    No  Female              0      No         No      13   
7263  9992-RRAMN   Yes    Male              0     Yes         No      22   
7264  9992-UJOEL    No    Male              0      No         No       2   
7265  9993-LHIEB    No    Male              0     Yes        Yes      67   
7266  9995-HOTOH    No    Male              0     Yes        Yes      63   

     PhoneService MultipleLines InternetService  ... OnlineBackup  \
0             Yes            No             DSL  ...          Yes   
1             Yes           Yes             DSL  ...           No   
2             Yes            No     Fiber optic  ...           No   
3             Yes            No     Fiber optic  ...          Yes   
4             Yes            No     Fiber optic  ...           No   
...           ...           ...             ...  ...          ...   
7262          Yes            No             DSL  ...           No   
7263          Yes           Yes     Fiber optic  ...           No   
7264          Yes            No             DSL  ...          Yes   
7265          Yes            No             DSL  ...           No   
7266           No            No             DSL  ...          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No         Yes         Yes              No        One year   
1                  No          No          No             Yes  Month-to-month   
2                 Yes          No          No              No  Month-to-month   
3                 Yes          No         Yes             Yes  Month-to-month   
4                  No         Yes         Yes              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7262               No         Yes          No              No        One year   
7263               No          No          No             Yes  Month-to-month   
7264               No          No          No              No  Month-to-month   
7265              Yes         Yes          No             Yes        Two year   
7266              Yes          No         Yes             Yes        Two year   

     PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0                 Yes      Mailed check           65.60          593.3  
1                  No      Mailed check           59.90          542.4  
2                 Yes  Electronic check           73.90         280.85  
3                 Yes  Electronic check           98.00        1237.85  
4                 Yes      Mailed check           83.90          267.4  
...               ...               ...             ...            ...  
7262               No      Mailed check           55.15          742.9  
7263              Yes  Electronic check           85.10         1873.7  
7264              Yes      Mailed check           50.30          92.75  
7265               No      Mailed check           67.85        4627.65  
7266               No  Electronic check           59.00         3707.6  

[7267 rows x 21 columns]

In [20]:
colunas = df_novo.columns

for col in colunas:
  print('---------------------------------------------')
  print(f'Os valores únicos da Coluna "{col}": \n{df_novo[col].unique()}')
  print('---------------------------------------------')

---------------------------------------------
Os valores únicos da Coluna "customerID": 
['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']
---------------------------------------------
---------------------------------------------
Os valores únicos da Coluna "Churn": 
['No' 'Yes' '']
---------------------------------------------
---------------------------------------------
Os valores únicos da Coluna "gender": 
['Female' 'Male']
---------------------------------------------
---------------------------------------------
Os valores únicos da Coluna "SeniorCitizen": 
[0 1]
---------------------------------------------
---------------------------------------------
Os valores únicos da Coluna "Partner": 
['Yes' 'No']
---------------------------------------------
---------------------------------------------
Os valores únicos da Coluna "Dependents": 
['Yes' 'No']
---------------------------------------------
---------------------------------------------
Os

In [21]:
df.head()

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0              Yes      Mailed check            65.6          593.3  
1               No      Mailed check            59.9          542.4  
2              Yes  Electronic check            73.9         280.85  
3              Yes  Electronic check            98.0        1237.85  
4              Yes      Mailed check            83.9          267.4  

[5 rows x 21 columns]

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


In [23]:
df_novo['Charges.Total'] = df_novo['Charges.Total'].astype(float)

Convertendo a coluna `Churn` para boolean

In [24]:
df_novo['Churn'] = df_novo['Churn'].str.replace('Yes','1').replace('No','0')

In [25]:
df_novo

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0     0002-ORFBO     0  Female              0     Yes        Yes       9   
1     0003-MKNFE     0    Male              0      No         No       9   
2     0004-TLHLJ     1    Male              0      No         No       4   
3     0011-IGKFF     1    Male              1     Yes         No      13   
4     0013-EXCHZ     1  Female              1     Yes         No       3   
...          ...   ...     ...            ...     ...        ...     ...   
7262  9987-LUTYD     0  Female              0      No         No      13   
7263  9992-RRAMN     1    Male              0     Yes         No      22   
7264  9992-UJOEL     0    Male              0      No         No       2   
7265  9993-LHIEB     0    Male              0     Yes        Yes      67   
7266  9995-HOTOH     0    Male              0     Yes        Yes      63   

     PhoneService MultipleLines InternetService  ... OnlineBackup  \
0             Yes            No             DSL  ...          Yes   
1             Yes           Yes             DSL  ...           No   
2             Yes            No     Fiber optic  ...           No   
3             Yes            No     Fiber optic  ...          Yes   
4             Yes            No     Fiber optic  ...           No   
...           ...           ...             ...  ...          ...   
7262          Yes            No             DSL  ...           No   
7263          Yes           Yes     Fiber optic  ...           No   
7264          Yes            No             DSL  ...          Yes   
7265          Yes            No             DSL  ...           No   
7266           No            No             DSL  ...          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No         Yes         Yes              No        One year   
1                  No          No          No             Yes  Month-to-month   
2                 Yes          No          No              No  Month-to-month   
3                 Yes          No         Yes             Yes  Month-to-month   
4                  No         Yes         Yes              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7262               No         Yes          No              No        One year   
7263               No          No          No             Yes  Month-to-month   
7264               No          No          No              No  Month-to-month   
7265              Yes         Yes          No             Yes        Two year   
7266              Yes          No         Yes             Yes        Two year   

     PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0                 Yes      Mailed check           65.60         593.30  
1                  No      Mailed check           59.90         542.40  
2                 Yes  Electronic check           73.90         280.85  
3                 Yes  Electronic check           98.00        1237.85  
4                 Yes      Mailed check           83.90         267.40  
...               ...               ...             ...            ...  
7262               No      Mailed check           55.15         742.90  
7263              Yes  Electronic check           85.10        1873.70  
7264              Yes      Mailed check           50.30          92.75  
7265               No      Mailed check           67.85        4627.65  
7266               No  Electronic check           59.00        3707.60  

[7267 rows x 21 columns]

Padronizando os Yes e No para 1 e 0

In [26]:
df_aux = df_novo.copy()

In [27]:
df_novo['MultipleLines'] = df_novo['MultipleLines'].str.replace('Yes','1').replace('No','0')
df_novo['OnlineSecurity'] = df_novo['OnlineSecurity'].str.replace('Yes','1').replace('No','0')
df_novo['OnlineBackup'] = df_novo['OnlineBackup'].str.replace('Yes','1').replace('No','0')
df_novo['DeviceProtection'] = df_novo['DeviceProtection'].str.replace('Yes','1').replace('No','0')
df_novo['TechSupport'] = df_novo['TechSupport'].str.replace('Yes','1').replace('No','0')
df_novo['StreamingTV'] = df_novo['StreamingTV'].str.replace('Yes','1').replace('No','0')
df_novo['StreamingMovies'] = df_novo['StreamingMovies'].str.replace('Yes','1').replace('No','0')

In [38]:
df_novo['Partner'] = df_novo['Partner'].str.replace('Yes','1').replace('No','0')
df_novo['Dependents'] = df_novo['Dependents'].str.replace('Yes','1').replace('No','0')
df_novo['PhoneService'] = df_novo['PhoneService'].str.replace('Yes','1').replace('No','0')
df_novo['PaperlessBilling'] = df_novo['PaperlessBilling'].str.replace('Yes','1').replace('No','0')

In [28]:
for i in range(len(df_novo['Churn'])):
    if str(df_novo['Churn'].iloc[i]).strip() == '':
      df_novo['Churn'].iloc[i] = 0

/tmp/ipython-input-28-120506457.py:3: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df_novo['Churn'].iloc[i] = 0
/tmp/ipython-input-28-120506457.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

In [29]:
df_novo['Churn'] = df_novo['Churn'].astype(int)

In [30]:
df_novo['MultipleLines'] = df_novo['MultipleLines'].astype(int)
df_novo['OnlineSecurity'] = df_novo['OnlineSecurity'].astype(int)
df_novo['OnlineBackup'] = df_novo['OnlineBackup'].astype(int)
df_novo['DeviceProtection'] = df_novo['DeviceProtection'].astype(int)
df_novo['TechSupport'] = df_novo['TechSupport'].astype(int)
df_novo['StreamingTV'] = df_novo['StreamingTV'].astype(int)
df_novo['StreamingMovies'] = df_novo['StreamingMovies'].astype(int)

In [39]:
df_novo['Partner'] = df_novo['Partner'].astype(int)
df_novo['Dependents'] = df_novo['Dependents'].astype(int)
df_novo['PhoneService'] = df_novo['PhoneService'].astype(int)
df_novo['PaperlessBilling'] = df_novo['PaperlessBilling'].astype(int)

In [40]:
df_novo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   int64  
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   int64  
 5   Dependents        7267 non-null   int64  
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   int64  
 8   MultipleLines     7267 non-null   int64  
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   int64  
 11  OnlineBackup      7267 non-null   int64  
 12  DeviceProtection  7267 non-null   int64  
 13  TechSupport       7267 non-null   int64  
 14  StreamingTV       7267 non-null   int64  
 15  StreamingMovies   7267 non-null   int64  
 16  Contract          7267 non-null   object 


In [41]:
df_novo

customerID  Churn  gender  SeniorCitizen  Partner  Dependents  tenure  \
0     0002-ORFBO      0  Female              0        1           1       9   
1     0003-MKNFE      0    Male              0        0           0       9   
2     0004-TLHLJ      1    Male              0        0           0       4   
3     0011-IGKFF      1    Male              1        1           0      13   
4     0013-EXCHZ      1  Female              1        1           0       3   
...          ...    ...     ...            ...      ...         ...     ...   
7262  9987-LUTYD      0  Female              0        0           0      13   
7263  9992-RRAMN      1    Male              0        1           0      22   
7264  9992-UJOEL      0    Male              0        0           0       2   
7265  9993-LHIEB      0    Male              0        1           1      67   
7266  9995-HOTOH      0    Male              0        1           1      63   

      PhoneService  MultipleLines InternetService  ...  OnlineBackup  \
0                1              0             DSL  ...             1   
1                1              1             DSL  ...             0   
2                1              0     Fiber optic  ...             0   
3                1              0     Fiber optic  ...             1   
4                1              0     Fiber optic  ...             0   
...            ...            ...             ...  ...           ...   
7262             1              0             DSL  ...             0   
7263             1              1     Fiber optic  ...             0   
7264             1              0             DSL  ...             1   
7265             1              0             DSL  ...             0   
7266             0              0             DSL  ...             1   

      DeviceProtection  TechSupport  StreamingTV  StreamingMovies  \
0                    0            1            1                0   
1                    0            0            0                1   
2                    1            0            0                0   
3                    1            0            1                1   
4                    0            1            1                0   
...                ...          ...          ...              ...   
7262                 0            1            0                0   
7263                 0            0            0                1   
7264                 0            0            0                0   
7265                 1            1            0                1   
7266                 1            0            1                1   

            Contract PaperlessBilling     PaymentMethod Charges.Monthly  \
0           One year                1      Mailed check           65.60   
1     Month-to-month                0      Mailed check           59.90   
2     Month-to-month                1  Electronic check           73.90   
3     Month-to-month                1  Electronic check           98.00   
4     Month-to-month                1      Mailed check           83.90   
...              ...              ...               ...             ...   
7262        One year                0      Mailed check           55.15   
7263  Month-to-month                1  Electronic check           85.10   
7264  Month-to-month                1      Mailed check           50.30   
7265        Two year                0      Mailed check           67.85   
7266        Two year                0  Electronic check           59.00   

      Charges.Total  
0            593.30  
1            542.40  
2            280.85  
3           1237.85  
4            267.40  
...             ...  
7262         742.90  
7263        1873.70  
7264          92.75  
7265        4627.65  
7266        3707.60  

[7267 rows x 21 columns]

Renomeando e traduzindo as colunas do meu DataFrame

In [47]:
colunas_traduzidas = [
    'ID_Cliente',                # customerID
    'Evasao',                    # Churn
    'Genero',                    # gender
    'Idoso',                     # SeniorCitizen
    'Possui_Parceiro',           # Partner
    'Possui_Dependentes',        # Dependents
    'Meses_De_Contrato',         # tenure
    'Servico_Telefonico',        # PhoneService
    'Multiplas_Linhas',          # MultipleLines
    'Servico_Internet',          # InternetService
    'Seguranca_Online',          # OnlineSecurity
    'Backup_Online',             # OnlineBackup
    'Protecao_Dispositivo',      # DeviceProtection
    'Suporte_Tecnico',           # TechSupport
    'Streaming_TV',              # StreamingTV
    'Streaming_Filmes',          # StreamingMovies
    'Tipo_Contrato',             # Contract
    'Fatura_Papel',              # PaperlessBilling
    'Metodo_Pagamento',          # PaymentMethod
    'Gasto_Mensal',              # Charges.Monthly
    'Gasto_Total'                # Charges.Total
]

for i in range(len(colunas_traduzidas)):
  df_novo = df_novo.rename(columns={colunas[i]:colunas_traduzidas[i]})

In [48]:
df_novo

ID_Cliente  Evasao  Genero  Idoso  Possui_Parceiro  Possui_Dependentes  \
0     0002-ORFBO       0  Female      0                1                   1   
1     0003-MKNFE       0    Male      0                0                   0   
2     0004-TLHLJ       1    Male      0                0                   0   
3     0011-IGKFF       1    Male      1                1                   0   
4     0013-EXCHZ       1  Female      1                1                   0   
...          ...     ...     ...    ...              ...                 ...   
7262  9987-LUTYD       0  Female      0                0                   0   
7263  9992-RRAMN       1    Male      0                1                   0   
7264  9992-UJOEL       0    Male      0                0                   0   
7265  9993-LHIEB       0    Male      0                1                   1   
7266  9995-HOTOH       0    Male      0                1                   1   

      Meses_De_Contrato  Servico_Telefonico  Multiplas_Linhas  \
0                     9                   1                 0   
1                     9                   1                 1   
2                     4                   1                 0   
3                    13                   1                 0   
4                     3                   1                 0   
...                 ...                 ...               ...   
7262                 13                   1                 0   
7263                 22                   1                 1   
7264                  2                   1                 0   
7265                 67                   1                 0   
7266                 63                   0                 0   

     Servico_Internet  ...  Backup_Online  Protecao_Dispositivo  \
0                 DSL  ...              1                     0   
1                 DSL  ...              0                     0   
2         Fiber optic  ...              0                     1   
3         Fiber optic  ...              1                     1   
4         Fiber optic  ...              0                     0   
...               ...  ...            ...                   ...   
7262              DSL  ...              0                     0   
7263      Fiber optic  ...              0                     0   
7264              DSL  ...              1                     0   
7265              DSL  ...              0                     1   
7266              DSL  ...              1                     1   

      Suporte_Tecnico  Streaming_TV  Streaming_Filmes   Tipo_Contrato  \
0                   1             1                 0        One year   
1                   0             0                 1  Month-to-month   
2                   0             0                 0  Month-to-month   
3                   0             1                 1  Month-to-month   
4                   1             1                 0  Month-to-month   
...               ...           ...               ...             ...   
7262                1             0                 0        One year   
7263                0             0                 1  Month-to-month   
7264                0             0                 0  Month-to-month   
7265                1             0                 1        Two year   
7266                0             1                 1        Two year   

     Fatura_Papel  Metodo_Pagamento Gasto_Mensal  Gasto_Total  
0               1      Mailed check        65.60       593.30  
1               0      Mailed check        59.90       542.40  
2               1  Electronic check        73.90       280.85  
3               1  Electronic check        98.00      1237.85  
4               1      Mailed check        83.90       267.40  
...           ...               ...          ...          ...  
7262            0      Mailed check        55.15       742.90  
7263            1  Electronic check        85.10      1873.70  
7264  

Criando a coluna Contas Diárias

In [51]:
df_novo['Gasto_Diario'] = round(df_novo['Gasto_Mensal'] / 30, 2)

In [52]:
df_novo.head()

ID_Cliente  Evasao  Genero  Idoso  Possui_Parceiro  Possui_Dependentes  \
0  0002-ORFBO       0  Female      0                1                   1   
1  0003-MKNFE       0    Male      0                0                   0   
2  0004-TLHLJ       1    Male      0                0                   0   
3  0011-IGKFF       1    Male      1                1                   0   
4  0013-EXCHZ       1  Female      1                1                   0   

   Meses_De_Contrato  Servico_Telefonico  Multiplas_Linhas Servico_Internet  \
0                  9                   1                 0              DSL   
1                  9                   1                 1              DSL   
2                  4                   1                 0      Fiber optic   
3                 13                   1                 0      Fiber optic   
4                  3                   1                 0      Fiber optic   

   ...  Protecao_Dispositivo  Suporte_Tecnico  Streaming_TV  Streaming_Filmes  \
0  ...                     0                1             1                 0   
1  ...                     0                0             0                 1   
2  ...                     1                0             0                 0   
3  ...                     1                0             1                 1   
4  ...                     0                1             1                 0   

    Tipo_Contrato  Fatura_Papel  Metodo_Pagamento  Gasto_Mensal Gasto_Total  \
0        One year             1      Mailed check          65.6      593.30   
1  Month-to-month             0      Mailed check          59.9      542.40   
2  Month-to-month             1  Electronic check          73.9      280.85   
3  Month-to-month             1  Electronic check          98.0     1237.85   
4  Month-to-month             1      Mailed check          83.9      267.40   

   Gasto_Diario  
0          2.19  
1          2.00  
2          2.46  
3          3.27  
4          2.80  

[5 rows x 22 columns]

In [53]:
df_novo.to_json('dados_tratados.json', index=False)

#📊 Carga e análise

## Distribuição da Evasão

In [54]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In [76]:
df_novo['Evasao'].value_counts()

Evasao
0    5398
1    1869
Name: count, dtype: int64

Gráfico da Distribuição da Evasão

In [77]:
labels = ['Não Evasão','Evasão']
values = df_novo['Evasao'].value_counts()

# Create subplots: use 'domain' type for Pie subplot
fig = go.Figure(data=[go.Pie(
    labels=labels,
    values=values,
    pull=[0, 0.01],
    name="Distribuição da Evasão")])

# Use `hole` to create a donut-like pie chart
fig.update_traces(
    hole=.4,
    hoverinfo="label+percent")

fig.update_layout(
    title_text="Distribuição da Evasão",
    title_font_size=20,
    title_font_weight='bold',
    title_x=0.5,
    # Add annotations in the center of the donut pies.
    annotations=[dict(text='Evasão',
                      x=0.5,
                      y=0.5,
                      font_size=20,
                      font_weight='bold',
                      showarrow=False,
                      xanchor="center")])

fig.show()

Com o seguinte gráfico conseguimos perceber que mais de 25,7% sairam da empresa e os 74,3% permanecem ainda na empresa. Essas percentagem correspondem a um número de 1869 clientes que sairam, em relação a um número de 5398 clientes que ainda permanecem na Empresa.

#📄Relatorio Final